File: Day-9/day9.hs
import Data.List 

type Coordinate = (Int, Int)

main = do  
        fileData <- parseFile <$> readFile "input.txt"
        let partA = solveInput (replicate 2 (0, 0)) $ fileData
        let partB = solveInput (replicate 10 (0, 0)) $ fileData
        print $ "Solution to part 1: " ++ show partA
        print $ "Solution to part 2: " ++ show partB

parseFile :: String -> [Coordinate]
parseFile = concatMap ((\[dir, steps] -> replicate (read steps) $ directions dir) . words) . lines

directions :: String -> Coordinate
directions "L" = (-1, 0)
directions "R" = (1, 0)
directions "U" = (0, 1)
directions "D" = (0, -1)

evalMove :: Coordinate -> Coordinate -> Coordinate
evalMove (x1, y1) (x2, y2) | (dx, dy) == (2, 2) = (ddx, ddy)
                           | dx == 2            = (ddx, y1)
                           | dy == 2            = (x1, ddy)
                           | otherwise          = (x2, y2)
                           where
                                (dx, dy) = (abs (x1 - x2), abs (y1 - y2))
                                ddx = if x1 > x2 then x1 - 1 else x1 + 1
                                ddy = if y1 > y2 then y1 - 1 else y1 + 1


moveDir :: [Coordinate] -> Coordinate -> [Coordinate]
moveDir xs (dy, dx) = foldl (\nk p -> nk ++ [evalMove (last nk) p]) [(nhx, nhy)] xy
                        where 
                             (hx, hy):xy = xs
                             (nhx, nhy) = (hx + dx, hy + dy)

solveInput :: [Coordinate] -> [Coordinate] -> Int 
solveInput start steps = length . nub . map last . scanl moveDir start $ steps

File: Day-9/day9.py
def evalMove(head, tail):
    dx = head[0] - tail[0]
    dy = head[1] - tail[1]

    ddx, ddy = tail

    if (abs(dx) == 2) and (abs(dy) == 2):
        tail[0] += dx//2
        tail[1] += dy//2
    else:
        if abs(dx) == 2:
            tail[0] += dx//2
            tail[1] = head[1]
        elif abs(dy) == 2:
            tail[1] += dy//2
            tail[0] = head[0]

    new_x, new_y = tail
    return ((new_x != ddx) or (new_y != ddy))

def solveInput(rope, steps):
    visited = set()
    tail = rope[-1]
    head = rope[0]
    visited.add(str(tail))

    for step in steps:
        dir, dst = step.split()

        for _ in range(int(dst)):
            if dir == 'U':
                head[1] += 1
            elif dir == 'D':
                head[1] -= 1
            elif dir == 'R':
                head[0] += 1
            else:
                head[0] -= 1

            motion = True

            for n in range(1, len(rope)):
                if motion:
                    motion = evalMove(rope[n-1], rope[n])

            if motion:
                visited.add(str(rope[-1]))

    return len(visited)

file = open("Day-9/input.txt", "r")
steps = [line.strip() for line in file.readlines()]
file.close()

ropeOne = [[0, 0], [0, 0]]
ropeTwo = [[0, 0] for _ in range(10)]

print("Solution to part 1: ", solveInput(ropeOne, steps))
print("Solution to part 2: ", solveInput(ropeTwo, steps))